sound/soc/sof/ipc4-loader.c

```diff
@@ -423,6 +423,62 @@ int sof_ipc4_reload_fw_libraries(struct snd_sof_dev *sdev)
 	return ret;
 }
 
+/**
+ * sof_ipc4_update_basecfg_from_manifest - Update the base config from manifest
+ * @sdev: SOF device
+ * @fw_module: pointer struct sof_ipc4_fw_module to parse
+ * @basecfg: Pointer to the base_config to update
+ *
+ * Note: currently only used to update the CPC value if it is not set from
+ * topology already.
+ */
+void sof_ipc4_update_basecfg_from_manifest(struct snd_sof_dev *sdev,
+					   struct sof_ipc4_fw_module *fw_module,
+					   struct sof_ipc4_base_module_cfg *basecfg)
+{
+	const struct sof_man4_module_config *fm_config;
+	u32 cpc_pick = 0;
+	u32 max_cpc = 0;
+	int i;
+
+	/* The CPC is set from topology, ignoring manifest */
+	if (basecfg->cpc)
+		return;
+
+	if (!fw_module || !fw_module->fm_config)
+		return;
+
+	fm_config = fw_module->fm_config;
+	for (i = 0; i < fw_module->man4_module_entry.cfg_count; i++) {
+		if (basecfg->obs == fm_config[i].obs &&
+		    basecfg->ibs == fm_config[i].ibs &&
+		    cpc_pick < fm_config[i].cpc)
+			cpc_pick = fm_config[i].cpc;
+
+		if (max_cpc < fm_config[i].cpc)
+			max_cpc = fm_config[i].cpc;
+	}
+
+	if (!cpc_pick) {
+		dev_dbg(sdev->dev,
+			"%s: No direct CPC match based on obs/ibs %u/%u\n",
+			fw_module->man4_module_entry.name, basecfg->obs, basecfg->ibs);
+		cpc_pick = max_cpc;
+	}
+
+	if (cpc_pick) {
+		dev_dbg(sdev->dev, "%s: CPC from manifest: %u\n",
+			fw_module->man4_module_entry.name, cpc_pick);
+	} else {
+		cpc_pick = basecfg->ibs * basecfg->obs / 1000;
```

Comment:
I am lost on why this formula would help? 

````
struct sof_ipc4_base_module_cfg {
	uint32_t cpc; /* the max count of Cycles Per Chunk processing */
	uint32_t ibs; /* input Buffer Size (in bytes)  */
	uint32_t obs; /* output Buffer Size (in bytes) */
	uint32_t is_pages; /* number of physical pages used */
	struct sof_ipc4_audio_format audio_fmt;
} __packed __aligned(4);
````

How would multiplying buffer sizes give us a valid number of cycles? That's super odd.

Reply:
I can leave this out and just send 0 to the firmware (as we do in some cases atm), but that is not good for the firmware.
If the module is missing this information then there is really no way that we can guess as each module can have different requirements. Based on the number I have seen around, this formula gives a number in the ballpark of measured numbers among different ones.
In general bigger the buffer, more processing is needed, but I cannot figure out better way atm.

Reply:
CPC is cycle per chunk. How is the 'chunk' related to the input and output buffers?

Undocumented assumptions left and right with all this. @ujfalusi @ranj063 @kv2019i @lgirdwood unless we are able to explain in layman terms what the cpc is and how it's determine I am not hot on submitting those patches upstream. When it looks like a hack, etc.

Reply:
@plbossart, I tend to agree, so let's leave it as 0 if it is not provided and think of another fallback path as a followup, which should be kernel only I'm afraid.